dev/bots/custom_rules/no_double_clamp.dart

```diff
@@ -0,0 +1,104 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:analyzer/dart/analysis/results.dart';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/dart/element/type.dart';
+import 'package:path/path.dart' as path;
+
+import '../utils.dart';
+import 'analyze.dart';
+
+/// Verify that we use clampDouble instead of double.clamp for performance
+/// reasons.
+///
+/// See also:
+///   * https://github.com/flutter/flutter/pull/103559
+///   * https://github.com/flutter/flutter/issues/103917
+final AnalyzeRule noDoubleClamp = _NoDoubleClamp();
+
+class _NoDoubleClamp implements AnalyzeRule {
+  final Map<ResolvedUnitResult, List<AstNode>> _errors = <ResolvedUnitResult, List<AstNode>>{};
+
+  @override
+  void reportViolations(String workingDirectory) {
+    if (_errors.isEmpty) {
+      return;
+    }
+
+    String locationInFile(ResolvedUnitResult unit, AstNode node) {
+      return '${path.relative(path.relative(unit.path, from: workingDirectory))}:${unit.lineInfo.getLocation(node.offset).lineNumber}';
+    }
+
+    foundError(<String>[
+      for (final MapEntry<ResolvedUnitResult, List<AstNode>> entry in _errors.entries)
+        for (final AstNode node in entry.value)
+          '${locationInFile(entry.key, node)}: ${node.parent}',
+      '\n${bold}For performance reasons, we use a custom "clampDouble" function instead of using "double.clamp".$reset',
+    ]);
+  }
+
+  @override
+  void applyTo(ResolvedUnitResult unit) {
```

Comment:
code organisation nit: Maybe place `applyTo` above `reportViolations` as per https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#order-other-class-members-in-a-way-that-makes-sense.